osu.Game.Rulesets.Catch/Skinning/LegacyComboCounter.cs

```diff
@@ -0,0 +1,162 @@
+                Origin = Anchor.Centre,
+            };
+        }
+
+        private IBindable<bool> isBreakTime;
+
+        [Resolved(canBeNull: true)]
+        private GameplayBeatmap beatmap { get; set; }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            isBreakTime = beatmap?.IsBreakTime.GetBoundCopy();
+            isBreakTime?.BindValueChanged(b =>
+            {
+                if (b.NewValue)
+                    this.FadeOut(400.0, Easing.OutQuint);
+            });
+        }
+
+        public void DisplayInitialCombo(int combo) => updateCombo(combo, null, true);
+        public void UpdateCombo(int combo, Color4? hitObjectColour) => updateCombo(combo, hitObjectColour, false);
+
+        private LegacyRollingCounter lastExplosion;
+
+        private void updateCombo(int combo, Color4? hitObjectColour, bool immediate)
+        {
+            // There may still be existing transforms to the counter (including value change after 250ms),
+            // finish them immediately before new transforms.
+            counter.FinishTransforms();
+
+            // Combo fell to zero, roll down and fade out the counter.
+            if (combo == 0)
+            {
+                counter.Current.Value = 0;
+                if (lastExplosion != null)
+                    lastExplosion.Current.Value = 0;
+
+                this.FadeOut(immediate ? 0.0 : 400.0, Easing.Out);
+                return;
+            }
+
+            // Remove last explosion to not conflict with the upcoming one.
+            if (lastExplosion != null)
+                RemoveInternal(lastExplosion);
+
+            this.FadeIn().Delay(1000.0).FadeOut(300.0);
+
+            // For simplicity, in the case of rewinding we'll just set the counter to the current combo value.
+            immediate |= Time.Elapsed < 0;
+
+            if (immediate)
+            {
+                counter.SetCountWithoutRolling(combo);
+                return;
+            }
+
+            counter.ScaleTo(1.5f).ScaleTo(0.8f, 250.0, Easing.Out)
+                   .OnComplete(c => c.SetCountWithoutRolling(combo));
+
+            counter.Delay(250.0).ScaleTo(1f).ScaleTo(1.1f, 60.0).Then().ScaleTo(1f, 30.0);
+
+            var explosion = new LegacyRollingCounter(skin, fontName, fontOverlap)
+            {
+                Alpha = 0.65f,
+                Anchor = Anchor.Centre,
+                Origin = Anchor.Centre,
+                Scale = new Vector2(1.5f),
+                Colour = hitObjectColour ?? Color4.White,
+                Depth = 1f,
+            };
+
+            AddInternal(explosion);
+
+            explosion.SetCountWithoutRolling(combo);
+            explosion.ScaleTo(1.9f, 400.0, Easing.Out)
+                     .FadeOut(400.0)
+                     .Expire(true);
+
+            lastExplosion = explosion;
+        }
+
+        private class LegacyRollingCounter : RollingCounter<int>
```

Comment:
not entirely sure about this being an inner class - I can see uses for this elsewhere (legacy counters in mania and taiko, for one)

Reply:
thought about that, but went with leaving it until legacy counters are implemented on the legacy HUD overlay, but sure can move this.

Reply:
actually the whole `LegacyComboCounter` thing seems to be movable outside, and makes sense to do so.

Reply:
i don't know about the entire thing - aren't the animations ruleset-specific? pretty sure other rulesets animate the counter differently in stable